source/adapters/level_zero/command_buffer.cpp

```diff
@@ -501,78 +588,132 @@ bool canBeInOrder(ur_context_handle_t Context,
              : false;
 }
 
+/**
+ * Append the initial barriers to the Compute and Copy command-lists.
+ * @param CommandBuffer The CommandBuffer
+ * @return UR_RESULT_SUCCESS or an error code on failure.
+ */
+ur_result_t appendExecutionWaits(ur_exp_command_buffer_handle_t CommandBuffer) {
+
+  std::vector<ze_event_handle_t> PrecondEvents;
+  if (CommandBuffer->ZeCommandListResetEvents) {
+    PrecondEvents.push_back(CommandBuffer->AllResetEvent->ZeEvent);
+  }
+  if (!CommandBuffer->UseImmediateAppendPath) {
+    PrecondEvents.push_back(CommandBuffer->WaitEvent->ZeEvent);
+  }
+
+  ZE2UR_CALL(zeCommandListAppendBarrier,
```

Comment:
Does this have to be a barrier? Can you use `zeCommandListAppendWaitOnEvents` instead?

Reply:
The L0 documentation is not very clear. Could you clarify what are the advantages of zeCommandListAppendWaitOnEvents  over a barrier with a nullptr signal event?

Does zeCommandListAppendWaitOnEvents  provide the same guarantees as a barrier? This is, does it guarantee that any commands appended after the wait are not executed before the events are signalled?

Reply:
>zeCommandListAppendWaitOnEvents over a barrier with a nullptr signal event?

Barrier is coarse-grain and enforces global memory ordering.

>Does zeCommandListAppendWaitOnEvents provide the same guarantees as a barrier? This is, does it guarantee that any commands appended after the wait are not executed before the events are signalled

Not for out-of-order cmdlists AFAIK. That is, commands can be reordered past `AppendWaitOnEvents` if there's no explicit event synchronization point between these two operations. But you are using events anyway, so that's why I've asked.

@nrspruit ping, in case I got anything wrong.